File: Chapter13/C#/Server Side/Order Processor System/Order Processor/OrderSystem.cs
using System;
using System.Drawing;
using System.Collections;

using Asap.Cars.CommonTypes;

namespace Asap.Cars.OrderProcessor
{
	/// <summary>
	/// OrderSystem is the sole server-side object that
	/// clients deal with. The class is derived from
	/// MarshalByRefObject to make it available as a
	/// distributed object using the .Net Remoting Framework
	/// </summary>
	public class OrderSystem: MarshalByRefObject
	{
    static int ordersReceived;
    static public int OrdersReceived
    {
      get {return ordersReceived;}
    }

		public OrderSystem()
		{
		}

    Router router;
    public Router Router
    {
      get
      {
        if (router == null)
          router = new Router();
        return router;
      }
    }

    public string[] GetModels()
    {
      // the following data would probably be fetched from a database
      string[] models = new string[] {"Model 1", "Model 2", "Model 3"};
      return models;
    }

    public string[] GetStyles(string theModel)
    {
      // the following data would probably be fetched from a database
      string[] styles = new string[] {"Style 1", "Style 2", "Style 3"};
      return styles;
    }

    public Color[] GetColors(string theModel, string theStyle)
    {
      // the following data would probably be fetched from a database
      Color[] colors = new Color[] {Color.White, Color.Navy, Color.Lavender};
      return colors;
    }

    // each entry in the returned array is a PricedItem[]
    public ArrayList GetOptions(string theModel, string theStyle)
    {
      // the following data would probably be fetched from a database
      ArrayList options = new ArrayList();

      ArrayList category1 = new ArrayList();
      PricedItem[] category1Items = new PricedItem[] {
        new PricedItem("Category 1", 0),
        new PricedItem("Option 1", 111),
        new PricedItem("Option 2", 222),
        new PricedItem("Option 3", 333)};
      options.Add(category1Items);

      ArrayList category2 = new ArrayList();
      PricedItem[] category2Items = new PricedItem[] {
        new PricedItem("Category 2", 0),
        new PricedItem("Option 11", 777),
        new PricedItem("Option 22", 888),
        new PricedItem("Option 33", 999)};
      options.Add(category2Items);

      return options;
    }

    public void SubmitOrder(string theModel, string theStyle, Color theColor, PricedItem[] theOptions)
    {
      WorkOrder workOrder = new WorkOrder(theModel, theStyle, theColor, theOptions);
      Router.SubmitOrder(workOrder);
      ordersReceived++;
    }

    public System.Messaging.Message[] GetAllQueuedMessages()
    {
      return Router.GetAllQueuedMessages();
    }

    public void ClearQueuedMessages()
    {
      Router.ClearQueuedMessages();
    }
  }
}

File: Chapter13/VB/Server Side/Order Processor System/Order Processor/OrderSystem.vb
Imports System.Drawing
Imports Asap.Cars.CommonTypes

' OrderSystem is the sole server-side object that
' clients deal with. The class is derived from
' MarshalByRefObject to make it available as a
' distributed object using the .Net Remoting Framework
Public Class OrderSystem
  Inherits MarshalByRefObject

  Private Shared _ordersReceived As Integer
  Public Shared ReadOnly Property OrdersReceived() As Integer
    Get
      Return _ordersReceived
    End Get
  End Property

  Public Sub New()
  End Sub

  Private _router As Router
  Public ReadOnly Property Router() As Router
    Get
      If _router Is Nothing Then
        _router = New Router
      End If
      Return _router
    End Get
  End Property

  Public Function GetModels() As String()
    ' the following data would probably be fetched from a database
    Dim models As String() = New String() {"Model 1", "Model 2", "Model 3"}
    Return models
  End Function

  Public Function GetStyles(ByVal theModel As String) As String()
    ' the following data would probably be fetched from a database
    Dim styles As String() = New String() {"Style 1", "Style 2", "Style 3"}
    Return styles
  End Function

  Public Function GetColors(ByVal theModel As String, ByVal theStyle As String) As Color()
    ' the following data would probably be fetched from a database
    Dim colors As Color() = New Color() {Color.White, Color.Navy, Color.Lavender}
    Return colors
  End Function

  ' each entry in the returned array is a PricedItem[]
  Public Function GetOptions(ByVal theModel As String, ByVal theStyle As String) As ArrayList
    ' the following data would probably be fetched from a database
    Dim options As New ArrayList

    Dim category1 As New ArrayList
    Dim category1Items As PricedItem() = New PricedItem() _
    {New PricedItem("Category 1", 0), _
      New PricedItem("Option 1", 111), _
      New PricedItem("Option 2", 222), _
      New PricedItem("Option 3", 333)}
    options.Add(category1Items)

    Dim category2 As New ArrayList
    Dim category2Items As PricedItem() = New PricedItem() _
    {New PricedItem("Category 2", 0), _
      New PricedItem("Option 11", 777), _
      New PricedItem("Option 22", 888), _
      New PricedItem("Option 33", 999)}
    options.Add(category2Items)

    Return options
  End Function

  Public Sub SubmitOrder(ByVal theModel As String, ByVal theStyle As String, ByVal theColor As Color, ByVal theOptions As PricedItem())
    Dim wo As New WorkOrder(theModel, theStyle, theColor, theOptions)
    Router.SubmitOrder(wo)
    _ordersReceived += 1
  End Sub

  Public Function GetAllQueuedMessages() As System.Messaging.Message()
    Return Router.GetAllQueuedMessages()
  End Function

  Public Sub ClearQueuedMessages()
    Router.ClearQueuedMessages()
  End Sub
End Class
